### PR TITLE
network: Fix memleaks found by test

### LINKED ISSUE #10149: There seem to be a bunch of memory leaks in systemd-networkd
I've just run the tests from https://github.com/systemd/systemd/pull/10147 and now I can see a lot of issues reported by Valgrind, which I'm going to paste here one by one.

```sh
==28238==
==28238== HEAP SUMMARY:
==28238==     in use at exit: 16,512 bytes in 4 blocks
==28238==   total heap usage: 4,071 allocs, 4,067 frees, 1,370,399 bytes allocated
==28238==
==28238== 128 bytes in 1 blocks are definitely lost in loss record 1 of 4
==28238==    at 0x4C2EA1E: calloc (vg_replace_malloc.c:711)
==28238==    by 0x12C69C: routing_policy_rule_new (networkd-routing-policy-rule.c:20)
==28238==    by 0x12D105: routing_policy_rule_add_internal (networkd-routing-policy-rule.c:236)
==28238==    by 0x12D32D: routing_policy_rule_add_foreign (networkd-routing-policy-rule.c:296)
==28238==    by 0x11CBE3: manager_rtnl_process_rule (networkd-manager.c:854)
==28238==    by 0x11F6FA: manager_rtnl_enumerate_rules (networkd-manager.c:1670)
==28238==    by 0x1169D1: main (networkd.c:114)
==28238==
==28238== LEAK SUMMARY:
==28238==    definitely lost: 128 bytes in 1 blocks
==28238==    indirectly lost: 0 bytes in 0 blocks
==28238==      possibly lost: 0 bytes in 0 blocks
==28238==    still reachable: 16,384 bytes in 3 blocks
==28238==         suppressed: 0 bytes in 0 blocks
==28238== Reachable blocks (those to which a pointer was found) are not shown.
==28238== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==28238==
```

### PR DESCRIPTION
**Pull request by yuwata**
Not tested, but I hope this fixes #10149.

cc @evverx.

### CODE REVIEW

**Comment by yuwata** [comment]
@evverx I have tried to fix the heap-buffer-overflow in #10149. But not tested. If possible, could you add a test for the overflow?

File: src/libsystemd-network/sd-dhcp6-client.c [hunk 1]
```diff
@@ -1091,7 +1090,6 @@ static int client_receive_advertise(sd_dhcp6_client *client, DHCP6Message *adver
 
         if (r < 0 || pref_advertise > pref_lease) {
                 client_set_lease(client, lease);
-                lease = NULL;
```
Comment (keszybz) [inline src/libsystemd-network/sd-dhcp6-client.c:1094]: This is correct.

Another way would be to make `client_set_lease` take ownership of the reference. I would prefer that.

**Comment by keszybz** [comment]
I pushed another trivial cleanup on top.

**Comment by yuwata** [comment]
Thanks.

**Comment by evverx** [comment]
Also, I'm not sure about fixing the bugs found by the fuzzer here because everything is just piling up without a reliable way to verify that it works, let alone regression testing. Ideally, it would be great to get the fuzzer working, but it turns out that it's much more work than I anticipated and I'm a little busy doing other stuff. Plus, I'm probably not the best person to write fuzzers for the code I'm not exactly familiar with :-)

**Comment by yuwata** [comment]
> another low-hanging fruit 

Did you find that with this PR applied? Or on current master?

**Comment by evverx** [comment]
> Did you find that with this PR applied? Or on current master?

I cherry-picked https://github.com/systemd/systemd/pull/10186/commits/23976cb92867007a1b9f9d52b9efd3f904472ed0 and https://github.com/systemd/systemd/pull/10186/commits/cb1bdeaf56852275e6b0dd1fba932bb174767f70 to see whether the first issue was gone. And it's gone, by the way.

> I guess it is also fixed by cb1bdea.

Unfortunately, it isn't. Basically, that's why I think it'd be better to wait until the fuzzer were functional because without any testing whatsoever it's kind of hard to tell whether the situation is improved or not.

**Comment by evverx** [comment]
I've just tried to run the test from #10147 and something went wrong. Apparently, they have changed and now I can't even get them working, so I have no idea whether the leaks are gone or not. Sorry. I'll try once more tomorrow.

**Comment by yuwata** [comment]
@evverx As commented at #10149 and the above by @keszybz, there still exist so many leaks (address, route, sd-device, sd-event...). But, at least, memleaks related to `routing_policy_rule` cannot be seen any more with this PR. I hope you can also observe that.

**Comment by evverx** [comment]
@yuwata I added a reproducer for the issue you'd fixed, a reproducer for another issue uncovered by the fuzzer, the fix for that issue, and the fuzzer itself in https://github.com/systemd/systemd/pull/10200. I'd be great if you could take a look.

**Pull request merged**